src/coreclr/src/jit/morph.cpp

```diff
@@ -13705,24 +13715,55 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                                 }
                             }
                         }
-                        // If the type of the IND (typ) is a "small int", and the type of the local has the
-                        // same width, then we can reduce to just the local variable -- it will be
-                        // correctly normalized, and signed/unsigned differences won't matter.
-                        //
-                        // The below transformation cannot be applied if the local var needs to be normalized on load.
-                        else if (varTypeIsSmall(typ) && (genTypeSize(lvaTable[lclNum].lvType) == genTypeSize(typ)) &&
-                                 !lvaTable[lclNum].lvNormalizeOnLoad())
+                        bool matchingTypes  = (lclType == typ);
```

Comment:
I think the new logic is not correct for `tree->OperIs(GT_LCL_FLD)`. If you have a field then `lclType` is the parent struct type, not the field type.

Reply:
I stepped through this just now in the debugger and the lclType was `TYP_STRUCT`,  typ was `TYP_REF` and  `foldAndReturnTemp` was false. The method is `SemaphoreSlim:WaitUntilCountOrTimeoutAsync`.  The temp typ[e get set correctly to TYP_REF  on line 13938:

```
                    temp->gtType      = tree->gtType;
                    foldAndReturnTemp = true;
```

Reply:
Also GT_LCL_FLD is the pessimized form that forces the local to be in memory and it will always use the sign/zero extending loads.

Reply:
Even if it eventually "works", won't it reject some folding because the sizes won't match for some `GT_LCL_FLD` cases?

Reply:
We never create a GT_IND with a typ of TYP_UINT:

See Importer.cpp, line 13478:
```
            case CEE_LDIND_U4:
            case CEE_LDIND_I4:
                lclTyp = TYP_INT;
```